### PR TITLE
Implemented XPath Functions (Spec 6.2 & Spec 6.3)

### PR DESCRIPTION
**Pull request by Abbasi-Waqas**
Added a function that encodes reserved characters in a string that is intended to be used in the path segment of a URI (encode-for-uri)

https://www.w3.org/TR/xpath-functions-31/#anyURI-functions (6.2 fn:encode-for-uri)

### CODE REVIEW

File: src/expressions/functions/builtInFunctions_string.ts [hunk 1]
```diff
@@ -828,7 +850,13 @@ export default {
 			namespaceURI: FUNCTIONS_NAMESPACE_URI,
 			returnType: 'xs:integer*',
 		},
-
+		{
+			argumentTypes: ['xs:string?'],
+			callFunction: fnEncodeForUri,
+			localName: 'encode-for-uri',
+			namespaceURI: FUNCTIONS_NAMESPACE_URI,
+			returnType: 'xs:string?',
```
Comment (bwrrp) [inline src/expressions/functions/builtInFunctions_string.ts:858]: Should that be just xs:string - I don't think this function ever returns an empty sequence?

File: test/assets/unrunnableTestCases.csv [hunk 2]
```diff
@@ -2385,7 +2385,8 @@ nscons-042,AssertionError: expected [Function] to throw error matching /XQDY0102
 nscons-043,AssertionError: expected [Function] to throw error matching /XPTY0004/ but got 'No selector counterpart for: computedNamespaceConstructor.'
 nscons-044,AssertionError: expected [Function] to throw error matching /XPTY0004/ but got 'No selector counterpart for: computedNamespaceConstructor.'
 Constr-comppi-space-1,AssertionError: Expected XPath processing-instruction pi {' text'} to resolve to the given XML. Expected <?pi  text?> to equal <?pi text?>
-Constr-comppi-space-2,AssertionError: Expected XPath processing-instruction pi {'&#x20;&#x0A;&#x0D;&#x09;text'} to resolve to the given XML. Expected <?pi   	text?> to equal <?pi text?>
+Constr-comppi-space-2,AssertionError: Expected XPath processing-instruction pi {'&#x20;&#x0A;&#x0D;&#x09;text'} to resolve to the given XML. Expected <?pi   
+	text?> to equal <?pi text?>
```
Comment (bwrrp) [inline test/assets/unrunnableTestCases.csv:2389]: How did this change suddenly introduce a newline here?

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 3]
```diff
@@ -509,4 +509,15 @@ describe('functions over strings', () => {
...
 			chai.assert.isFalse(evaluateXPathToBoolean('codepoint-equal("a","b")', documentNode)));
 	});
+
+	describe('encode-for-uri()', () => {
+		it('Returns the correct example for the example in the specs', () => {
+			chai.assert.equal(evaluateXPathToString('encode-for-uri("http://www.example.com/00/Weather/CA/Los%20Angeles#ocean")'), "http%3A%2F%2Fwww.example.com%2F00%2FWeather%2FCA%2FLos%2520Angeles%23ocean")
+
+		})
+		
+		it('Returns empty for an empty sequence', () => {
```
Comment (mehmetscoskun) [inline test/specs/parsing/functions/functions.string.tests.ts:519]: If $uri-part is the empty sequence, the function returns the zero-length string.
fn:empty("") returns false().

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 4]
```diff
@@ -512,12 +512,16 @@ describe('functions over strings', () => {
...
+			chai.assert.equal(
+				evaluateXPathToString(
+					'encode-for-uri("http://www.example.com/00/Weather/CA/Los%20Angeles#ocean")'
+				),
+				'http%3A%2F%2Fwww.example.com%2F00%2FWeather%2FCA%2FLos%2520Angeles%23ocean'
+			);
+		});
+
+		it('If $uri-part is the empty sequence, the function returns the zero-length string.', () => {
+			chai.assert.equal(evaluateXPathToBoolean('empty(encode-for-uri(()))'), "");
```
Comment (mehmetscoskun) [inline test/specs/parsing/functions/functions.string.tests.ts:524]: I guess you should remove `empty` function:
chai.assert.isTrue(evaluateXPathToBoolean('encode-for-uri(()) = ""'));

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 5]
```diff
@@ -512,12 +512,16 @@ describe('functions over strings', () => {
...
-	})
+			chai.assert.equal(
+				evaluateXPathToString(
+					'encode-for-uri("http://www.example.com/00/Weather/CA/Los%20Angeles#ocean")'
+				),
+				'http%3A%2F%2Fwww.example.com%2F00%2FWeather%2FCA%2FLos%2520Angeles%23ocean'
+			);
+		});
+
+		it('If $uri-part is the empty sequence, the function returns the zero-length string.', () => {
```
Comment (mehmetscoskun) [inline test/specs/parsing/functions/functions.string.tests.ts:523]: For the consistency, I would write the description like this:
it('Returns the zero-length string if the uri-part is the empty sequence')

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 6]
```diff
@@ -520,8 +520,60 @@ describe('functions over strings', () => {
...
+	describe.only('iri-to-uri()', () => {
+		it('Returns the correct example for the example in the specs', () => {
+			chai.assert.equal(
+				evaluateXPathToString('iri-to-uri("http://www.example.com/~bébé")'),
+				'http://www.example.com/~b%C3%A9b%C3%A9'
+			);
+		});
+
+		it('Returns the zero-length string if the uri-part is the empty sequence', () => {
+			chai.assert.equal(evaluateXPathToString('iri-to-uri(())'), '');
```
Comment (mehmetscoskun) [inline test/specs/parsing/functions/functions.string.tests.ts:537]: Please check the comment for `encode-for-uri` about this.

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 7]
```diff
@@ -520,8 +520,60 @@ describe('functions over strings', () => {
...
+
+	describe.only('iri-to-uri()', () => {
+		it('Returns the correct example for the example in the specs', () => {
+			chai.assert.equal(
+				evaluateXPathToString('iri-to-uri("http://www.example.com/~bébé")'),
+				'http://www.example.com/~b%C3%A9b%C3%A9'
+			);
+		});
+
+		it('Returns the zero-length string if the uri-part is the empty sequence', () => {
```
Comment (mehmetscoskun) [inline test/specs/parsing/functions/functions.string.tests.ts:536]: `the iri-part` ?

File: test/specs/parsing/functions/functions.string.tests.ts [hunk 8]
```diff
@@ -520,8 +520,60 @@ describe('functions over strings', () => {
...
+			chai.assert.equal(evaluateXPathToString('iri-to-uri(())'), '');
+		});
+
+		it('qt3tests 1', () => {
```
[the remark this answers is not in the document]
Comment (mehmetscoskun) [reply test/specs/parsing/functions/functions.string.tests.ts:540]: Just I am curious. If these tests are in qt3tests, why do we need to have them in unittests as well?

[the remark this answers is not in the document]
Comment (Abbasi-Waqas) [reply test/specs/parsing/functions/functions.string.tests.ts:540]: These were failing tests, added them for testing purposes, have removed them now

**Review by DrRataplan — APPROVED** [review]

**Pull request merged**